### PR TITLE
Refactor GitHub steps into an Azure Wizard

### PR DESCRIPTION
**Pull request by nturinski**
Fixes https://github.com/microsoft/vscode-azureappservice/issues/764

### CODE REVIEW

File: appservice/src/github/GitHubBranchListStep.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        } while (!branchQuickPick);
+
+        context.branchData = branchQuickPick;
+    }
+
+    public shouldPrompt(context: IConnectToGitHubWizardContext): boolean {
+        return !context.branchData;
+    }
+
+    private async getBranches(context: IConnectToGitHubWizardContext, picksCache: ICachedQuickPicks<gitHubBranchData>): Promise<IAzureQuickPickItem<gitHubBranchData | undefined>[]> {
```
Comment (EricJizbaMSFT) [inline appservice/src/github/GitHubBranchListStep.ts]: Can we rename this to getBranchPicks?

File: appservice/src/github/GitHubBranchListStep.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+import { AzureWizardPromptStep, IAzureQuickPickItem } from 'vscode-azureextensionui';
+import { ext } from '../extensionVariables';
+import { nonNullProp } from '../utils/nonNull';
+import { getGitHubQuickPicksWithLoadMore, gitHubBranchData, gitHubWebResource, ICachedQuickPicks } from './connectToGitHub';
+import { IConnectToGitHubWizardContext } from './IConnectToGitHubWizardContext';
+
+export class GitHubBranchListStep extends AzureWizardPromptStep<IConnectToGitHubWizardContext> {
+    public async prompt(context: IConnectToGitHubWizardContext): Promise<void> {
+        const placeHolder: string = 'Choose branch';
```
Comment (EricJizbaMSFT) [inline appservice/src/github/GitHubBranchListStep.ts]: "branchQuickPick" seems unnecessary and misleading since it's the data, not the pick

```
do {
    context.branchData = (await ext.ui.showQuickPick(this.getBranches(context, picksCache), { placeHolder })).data;
} while (!context.branchData);
```

File: appservice/src/github/connectToGitHub.ts [hunk 3]
```diff
@@ -7,29 +7,35 @@ import { SiteSourceControl } from 'azure-arm-website/lib/models';
...
 // tslint:disable-next-line:no-reserved-keywords
-type gitHubWebResource = WebResource & { resolveWithFullResponse?: boolean, nextLink?: string, type?: string };
+export type gitHubWebResource = WebResource & { resolveWithFullResponse?: boolean, nextLink?: string, type?: string };
 
-export async function connectToGitHub(node: AzureTreeItem, client: SiteClient, context: IActionContext): Promise<void> {
-    const requestOptions: gitHubWebResource = new WebResource();
-    requestOptions.resolveWithFullResponse = true;
-    requestOptions.headers = {
+export async function connectToGitHub(node: AzureTreeItem, client: SiteClient, context: IConnectToGitHubWizardContext): Promise<void> {
+    context.requestOption = new WebResource();
```
Comment (EricJizbaMSFT) [inline appservice/src/github/connectToGitHub.ts]: Not a huge fan of re-using the same requestOption. Can you create a new `requestOption` inside `getGitHubJsonResponse` and take the url as a parameter instead?

Specifically worried about one of the requests setting a header or something and other requests inheriting that on accident

File: appservice/src/github/connectToGitHub.ts [hunk 4]
```diff
@@ -38,35 +44,39 @@ export async function connectToGitHub(node: AzureTreeItem, client: SiteClient, c
...
+            new GitHubRepoListStep(),
+            new GitHubBranchListStep()
+        ]
+    });
+
+    try {
+        await wizard.prompt();
+    } catch (error) {
+        const parsedError: IParsedError = parseError(error);
+        if (parsedError.message.indexOf('Bad credentials') > -1) {
```
Comment (EricJizbaMSFT) [inline appservice/src/github/connectToGitHub.ts]: Why did you move this logic outside getGitHubJsonResponse? If you ever change this to be a subwizard instead of it's own wizard - you'd have to move this logic back since you wouldn't call wizard.prompt() directly.

Comment (nturinski) [reply appservice/src/github/connectToGitHub.ts]: Well, it was because it was the only reason each step required the node and client to be passed in which seemed excessive.

Comment (EricJizbaMSFT) [reply appservice/src/github/connectToGitHub.ts]: Ah well I'm fine with the change in that case (although like I said you might have to refactor if you ever move this to a subwizard)

Comment (nturinski) [reply appservice/src/github/connectToGitHub.ts]: Future Nathan's problem 😼

File: appservice/src/github/connectToGitHub.ts [hunk 5]
```diff
@@ -98,7 +85,7 @@ async function showGitHubAuthPrompt(node: AzureTreeItem, client: SiteClient, con
     const goToPortal: vscode.MessageItem = { title: localize('goToPortal', 'Go to Portal') };
     let input: vscode.MessageItem | undefined = DialogResponses.learnMore;
     while (input === DialogResponses.learnMore) {
-        input = await vscode.window.showErrorMessage(invalidToken, goToPortal, DialogResponses.learnMore);
+        input = await vscode.window.showErrorMessage(invalidToken, { modal: true }, goToPortal, DialogResponses.learnMore);
```
Comment (nturinski) [inline appservice/src/github/connectToGitHub.ts]: I made this a modal because the Organization quick pick menu looked like it was still loading when there's an error until you click the dialog which could be confusing.

**Comment by nturinski** [comment]
I'm going to continue making changes to this since it'll be necessary for another PR. It shouldn't be anything huge, just passing in the name of org/repo like we talked about offline.

**Comment by nturinski** [comment]
Never mind, Matt didn't like the automatically getting org/repo so not gonna add more to this PR.
